File: src/main/java/com/treutec/plainjson/SimpleProducer.java
package com.treutec.plainjson;

import com.fasterxml.jackson.core.JsonProcessingException;
import com.github.javafaker.Faker;
import com.treutec.Constants;
import com.treutec.Person;
import java.util.Properties;
import java.util.concurrent.ExecutionException;
import java.util.concurrent.Future;
import org.apache.kafka.clients.producer.KafkaProducer;
import org.apache.kafka.clients.producer.Producer;
import org.apache.kafka.clients.producer.ProducerRecord;
import org.apache.kafka.common.serialization.StringSerializer;

public final class SimpleProducer {
  private final Producer<String, String> producer;

  public SimpleProducer(String brokers) {
    Properties props = new Properties();
    props.put("bootstrap.servers", brokers);
    props.put("key.serializer", StringSerializer.class);
    props.put("value.serializer", StringSerializer.class);
    producer = new KafkaProducer<>(props);
  }

  public void produce(int ratePerSecond) {
    long waitTimeBetweenIterationsMs = 1000L / (long)ratePerSecond;
    Faker faker = new Faker();

    while(true) {
      Person fakePerson = new Person(
          faker.name().firstName(),
          faker.name().lastName(),
          faker.date().birthday(),
          faker.address().city(),
          faker.internet().ipV4Address()
      );
      String fakePersonJson = null;
      try {
        fakePersonJson = Constants.getJsonMapper().writeValueAsString(fakePerson);
      } catch (JsonProcessingException e) {
        e.printStackTrace();
      }
      Future futureResult = producer.send(new ProducerRecord<>(Constants.getPersonsTopic(), fakePersonJson));
      try {
        Thread.sleep(waitTimeBetweenIterationsMs);
        futureResult.get();
      } catch (InterruptedException | ExecutionException e) {
        e.printStackTrace();
      }
    }
  }

  public static void main(String[] args) {
    new SimpleProducer("localhost:9092").produce(2);
  }
}

File: src/main/kotlin/com/treutec/plain/SimpleProducer.kt
package com.treutec.plain

import com.github.javafaker.Faker
import com.treutec.Person
import com.treutec.jsonMapper
import com.treutec.personsTopic
import org.apache.kafka.clients.producer.KafkaProducer
import org.apache.kafka.clients.producer.Producer
import org.apache.kafka.clients.producer.ProducerRecord
import org.apache.kafka.common.serialization.StringSerializer
import org.apache.log4j.LogManager
import java.util.Properties

// kafka-topics --zookeeper localhost:2181 --create --topic persons --replication-factor 1 --partitions 4
// kafka-console-consumer --bootstrap-server localhost:9092 --topic persons --property print.key=true

fun main(args: Array<String>) {
    SimpleProducer("localhost:9092").produce(2)
}

class SimpleProducer(brokers: String) {

    private val logger = LogManager.getLogger(javaClass)
    private val producer = createProducer(brokers)

    private fun createProducer(brokers: String): Producer<String, String> {
        val props = Properties()
        props["bootstrap.servers"] = brokers
        props["key.serializer"] = StringSerializer::class.java
        props["value.serializer"] = StringSerializer::class.java
        return KafkaProducer<String, String>(props)
    }

    fun produce(ratePerSecond: Int) {
        val waitTimeBetweenIterationsMs = 1000L / ratePerSecond
        logger.info("Producing $ratePerSecond records per second (1 every ${waitTimeBetweenIterationsMs}ms)")

        val faker = Faker()
        while (true) {
            val fakePerson = Person(
                    firstName = faker.name().firstName(),
                    lastName = faker.name().lastName(),
                    birthDate = faker.date().birthday(),
                    city = faker.address().city(),
                    ipAddress = faker.internet().ipV4Address()
            )
            logger.info("Generated a person: $fakePerson")

            val fakePersonJson = jsonMapper.writeValueAsString(fakePerson)
            logger.debug("JSON data: $fakePersonJson")

            val futureResult = producer.send(ProducerRecord(personsTopic, fakePersonJson))
            logger.debug("Sent a record")

            Thread.sleep(waitTimeBetweenIterationsMs)

            // wait for the write acknowledgment
            futureResult.get()
        }
    }
}
